Bugfix: Implement roll control on Solo gimbal

igorsgcampos opened the pull request:
I was testing https://github.com/ArduPilot/ardupilot/pull/15878 and noticed that Solo was not changing the roll angle of the gimbal according to what was commanded. Upon inspecting the code, I saw that it just ignored the roll input, despite the vehicle's capability of processing the command.

This fixes the lack of roll control for Solo, making it compliant with the mavlink DO_MOUNT_CONTROL specification.

igorsgcampos commented:
I asked to put this on the dev-call, but it was forgotten, I would appreciate if it could get reviewed, it is pretty straight-forward.

khancyr reviewed (approved):
I don't have SOLO, so I cannot check. 
The code is mostly a duplicate from the tilt code except on L399 where the angle limit are adapted to roll.

khancyr commented:
@igorsgcampos thanks for this, it looks good to me. Could you fix the commit message ?
it should be `AP_Mount: Implement roll control on Solo gimbal` 
I have tag this for the next devcall so it should be merged fast.

igorsgcampos commented:
Done, it is updated now!

khancyr commented:
Thanks, hopefully it should be merged on next devcall.

rmackay9 commented:
@igorsgcampos,

You've tested this change on a Solo gimbal?  The solo gimbal has very limited roll movement I think so we really should test it on real hardware to make sure that the change is useful.

igorsgcampos commented:
Yes, I've tested it. The specs claim 40 degrees, that is why I limited to 30.

tridge commented:
I've asked @Pedals2Paddles and @davidbuzz to check this

rmackay9 commented:
This looks good to me.

davidbuzz commented:
I have no particular opinion on this, but if its been tested on a real solo gimbal, it sounds ok it theory.   lets see what @Pedals2Paddles  says.

Pedals2Paddles commented:
I should be able to test this week. What is the functional purpose of this change? It says bug fix but I'm not aware of very being a bug requiring this? In other words what problem are we solving here?

peterbarker commented:
On Mon, 11 Jan 2021, Matt wrote:

> I should be able to test this week. What is the functional purpose of this change? It says bug fix but I'm not aware of very being a bug requiring this?
> In other words what problem are we solving here?


Not a bug fix.

With the apprioriate changes made you should be able to deliverately roll 
the camera.

igorsgcampos commented:
> I should be able to test this week. What is the functional purpose of this change? It says bug fix but I'm not aware of very being a bug requiring this? In other words what problem are we solving here?

The purpose is enabling #15878 to work.

It is a bugfix in the sense that presently, if the Solo is issued a DO_MOUNT_CONTROL, it will ignore the roll angle, thus presenting a different behavior than the protocol specification. Sorry if the description was confusing according to your standards, but I considered this deviation from the protocol a bug.

igorsgcampos commented:
Here is a log from a flight using this feature:

[log_84_2020-12-20-16-08-56.zip](https://github.com/ArduPilot/ardupilot/files/5802437/log_84_2020-12-20-16-08-56.zip)

igorsgcampos commented:
This overview of part of the images it took on the flight should demonstrate the behavior of the camera changing its roll:

![image](https://user-images.githubusercontent.com/22230664/104323257-56b1a680-54bc-11eb-8bcf-a7ccd2211252.png)

Pedals2Paddles commented:
Ah ok. That's pretty cool.  I believe you're probably the first person who's ever actually *wanted* Solo gimbal roll control outside of automated level stabilization.  Or at least the first to say something anyway.  I would consider this an enhancement rather than a bugfix actually.

I'm not equipped to test surveys like you are, nor do I have experience with it.  So I will defer to your expertise and abilities on it doing what you're trying to achieve there. Is the roll angle something your sending in a command from the survey mission?  When the mission is done or you're otherwise just wanting it to be level, does it go back normal operation remaining level?

igorsgcampos commented:
>Is the roll angle something your sending in a command from the survey mission?

Right now I implemented a new command, which will eventually move from AP_Camera to AP_Vehicle. The implementation picks the AP_Mount singleton and instructs the camera to roll left and right according to the command parameters. For reference:
https://github.com/mavlink/mavlink/pull/1508

>When the mission is done or you're otherwise just wanting it to be level, does it go back normal operation remaining level?

At the end, it remains at the last position, but I changed QGC so that it adds a trailing DO_MOUNT_CONTROL with orientations 0,0,0, thus making the camera point forward.

Pedals2Paddles commented:
Understood. IIRC, nothing else is sending it roll commands so the general users with a controller flying around with just the pitch control will not see any difference?  If so, then I am happy with your testing/implementation and will approve the PR review.

igorsgcampos commented:
> Understood. IIRC, nothing else is sending it roll commands so the general users with a controller flying around with just the pitch control will not see any difference? If so, then I am happy with your testing/implementation and will approve the PR review.

That is correct, it should not affect present functions, only enable roll control, which is currently unused.

Pedals2Paddles reviewed (approved):
Fantastic.  Nice work and thank you for the contribution back to the community with this.  I approve. @davidbuzz can you approve your review too if you do?

peterbarker commented:
Merged, thanks @igorsgcampos 

On the "bugfix" thing - we deviate from the MAVLink standard a *lot*, because the standard is silly in many places.... so that deviation isn't necessarily the bug :-)

davidbuzz reviewed (commented):
Ok

igorsgcampos commented:
> On the "bugfix" thing - we deviate from the MAVLink standard a _lot_, because the standard is silly in many places.... so that deviation isn't necessarily the bug :-)

I didn't mean the MAVLink protocol in this case, but rather ArduPilot's documentation: https://ardupilot.org/copter/docs/mission-command-list.html#do-mount-control

In this specific case, Solo was behaving different from other Copters, but I do understand that this might be a corner case and unlikely to affect the vast majority of users.

Thanks for merging this!
